istio/src/main/scala/io/buoyant/k8s/istio/ClusterCache.scala

```diff
@@ -3,8 +3,13 @@ package io.buoyant.k8s.istio
 import com.twitter.logging.Logger
 import com.twitter.util.{Closable, Future, Time}
 
-class ClusterCache(client: DiscoveryClient) extends Closable {
-  import ClusterCache._
+case class Cluster(dest: String, port: String)
+
+trait ClusterCache extends Closable {
```

Comment:
I assume this is factored this way for test-ability.  

Instantiating a ClusterCacheBackedByApi with a test client that returns mock data would test more of the codepath than implementing a test ClusterCache.  But it would also be a lot less ergonomic.  I just wanted to bring it up so we can consider the tradeoff.

Reply:
Yeah, I'm not really a big fan of the `trait X`/`class XBackedByApi` pattern that shows up a couple times in this PR. Elsewhere in Linkerd, when we need to mock an external API for testing, I think the general pattern is to inject a mock API with overrides, like [this](https://github.com/linkerd/linkerd/blob/80b320244571ca3f66ae4673de05a438c7e02def/interpreter/k8s/src/test/scala/io/buoyant/interpreter/k8s/ConfigMapInterpreterTest.scala#L163-L167). Like Alex said, this would exercise more of the implementation, and I think it also results in a less complex non-test codebase?

Reply:
Interesting, the idea behind the use of an interface was to increase ergonomics. My thinking here is that the usual pattern of implementing a fake `Service[Req,Resp]` on tests have a few drawbacks:
1) I think that it leaks a lot of "infrastructure" types into higher level tests. These tests shouldn't have any reason to even know about Finagle or any other piece of infrastructure
2) Maybe more important, I also think that it also creates coupling between the consumer of a type and its implementation. For example, consider https://github.com/linkerd/linkerd/pull/1650/files#diff-bcd0445a689106c21b39b7304f665801R36 . If we follow the usual pattern of implementing a fake `Service[Req,Resp]` there then the test for `InternalTrafficIdentifier` would make assumptions about the API used by `ClusterCacheBackedByApi`. If the next version of the API used by this class changes the URL or any other implementation detail we would have to change not only the code and tests for `ClusterCacheBackedByApi` but also the tests for `InternalTrafficIdentifier`. This feels like accidental coupling.
3) I am not sure if I agree that this would exercise the codebase more, at least in any useful manner. One rule of thumb that I like to follow when it comes to testing is _"don't mock what you don't own"_, and in this case I think that we might be trying to use unit tests for what should be integration tests. Instead, I would suggest that we keep as much as possible of our own logic independent from these dependencies (which drives back to the 1st point) and use integration tests for those instead of mocking APIs that aren't under our control.
4) The concept of ergonomics or complex code that we are discussing here are subjective, but I have the impression that one of the reasons for what I perceive to be low test coverage around these boundary classes is because it is actually a lot of work to have those fake `Service[Req,Resp]`, so much so that we tend to have a single "fixture" reused across multiple tests, which requires tests to use special requests to know which part of the fixture will be hit.

But with all that said, Linkerd is an existing code base and I am happy to comply with the existing testing strategy if we collectively think that's the way to go.